code/datums/components/proximity_monitor.dm

```diff
@@ -1,139 +1,512 @@
 
 /datum/component/proximity_monitor/RegisterWithParent()
-	. = ..()
-	if(ismovable(parent))
-		RegisterSignal(parent, COMSIG_MOVABLE_MOVED, .proc/HandleMove)
+	if(ismovable(hasprox_receiver))
+		RegisterSignal(hasprox_receiver, COMSIG_MOVABLE_MOVED, .proc/on_receiver_move)
+		RegisterSignal(hasprox_receiver, COMSIG_MOVABLE_DISPOSING, .proc/on_disposal_enter)
+		RegisterSignal(hasprox_receiver, COMSIG_MOVABLE_EXIT_DISPOSALS, .proc/on_disposal_exit)
+	map_nested_locs()
 
 /datum/component/proximity_monitor/UnregisterFromParent()
-	. = ..()
-	if(ismovable(parent))
-		UnregisterSignal(parent, COMSIG_MOVABLE_MOVED)
+	if(ismovable(hasprox_receiver))
+		UnregisterSignal(hasprox_receiver, list(COMSIG_MOVABLE_MOVED, COMSIG_MOVABLE_DISPOSING, COMSIG_MOVABLE_EXIT_DISPOSALS))
+	clear_nested_locs()
 
 /**
- * Called when the `parent` receives the `COMSIG_MOVABLE_MOVED` signal, which occurs when it `Move()`s
+ * Called when the `hasprox_receiver` moves.
  *
- * Code is only ran when there is no `Dir`, which occurs when the parent is teleported, gets placed into a storage item, dropped, or picked up.
- * Normal movement, for example moving 1 tile to the west, is handled by the `proximity_checker` objects.
+ * Arguments:
+ * * datum/source - this will be the `hasprox_receiver`
+ * * atom/old_loc - the location the receiver just moved from
+ * * dir - the direction the reciever just moved in
+ */
+/datum/component/proximity_monitor/proc/on_receiver_move(datum/source, atom/old_loc, dir)
+	SIGNAL_HANDLER
+
+	// It was just a normal tile-based move, so we return here.
+	if(dir)
+		move_prox_checkers(dir)
+		return
+
+	// Moving onto a turf.
+	if(!isturf(old_loc) && isturf(hasprox_receiver.loc))
+		toggle_checkers(TRUE)
+		clear_nested_locs()
+
+	// Moving into an object.
+	else if(always_active)
+		toggle_checkers(TRUE)
+		map_nested_locs()
+
+	// The receiver moved into something, but isn't `always_active`, so deactivate the checkers.
+	else
+		toggle_checkers(FALSE)
+
+	recenter_prox_checkers()
+
+/**
+ * Called when an atom in `nested_receiver_locs` list moves, if one exists.
  *
  * Arguments:
- * * source - this will be the `parent`
- * * OldLoc - the location the parent just moved from
- * * Dir - the direction the parent just moved in
- * * forced - if we were forced to move
+ * * atom/moved_atom - one of the atoms in `nested_receiver_locs`
+ * * atom/old_loc - the location `moved_atom` just moved from
+ * * dir - the direction `moved_atom` just moved in
  */
-/datum/component/proximity_monitor/proc/HandleMove(datum/source, atom/OldLoc, Dir, forced)
-	if(!Dir) // No dir means the parent teleported, or moved in a non-standard way like getting placed into disposals, onto a table, dropped, picked up, etc.
-		recenter_prox_checkers()
+/datum/component/proximity_monitor/proc/on_nested_loc_move(atom/moved_atom, atom/old_loc, dir)
+	SIGNAL_HANDLER
+
+	// It was just a normal tile-based move, so we return here.
+	if(dir)
+		move_prox_checkers(dir)
+		return
```

Comment:
I'm a little concerned that relying on `dir` might be risky, I think a few things don't necessarily set it when they should (beyond setting .loc). In concept I think it's good, but I'd be a little wary.

Reply:
Yeah it is a little risky. Currently though, the base component (what's on live) has been using this method for over a year and I've yet to hear about a bug like this. Hopefully that means there are not a lot of direct `loc` sets or moves to nullspace that are a problem. That or no one's reported it.

I think the best idea might be to just address those cases when they come up.

Reply:
Nope, this is based on the direction they moved in, and not in mob facing direction.